Fix a memory leak of EndpointReconciler

masap opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
The endpoint entry is created in ```etcd``` when the ```EndpointReconciler``` is created, but even if the ```EndpointReconciler``` is destroyed, the endpoint entry remains. This PR modifies the endpoint entry to be deleted by calling ```destroyFunc```.

The endpoint entry in ```etcd``` is allocated at here.
https://github.com/kubernetes/kubernetes/blob/e3de62298a730415c5d2ab72607ef6adadd6304d/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go#L141

This issue could occur when running this integration test.
```
$ make test-integration WHAT=./test/integration/apiserver/apply
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

fedebongio commented:
/assign @roycaihw 
(looks like this is the old version of what you've been working on lately, thanks!)
/triage accepted

masap commented:
/ok-to-test

masap committed:
Fix a memory leak of EndpointReconciler

The endpoint entity is created in etcd when the EndpointReconciler is created,
but even if the EndpointReconciler is destroyed, the endpoint entity remains.
This PR modifies the endpoint entity to be deleted by calling destroyFunc.

This issue could occur when running this integration test.
$ make test-integration WHAT=./test/integration/apiserver/apply

masap commented:
rebased

masap commented:
@roycaihw Is there any comment?

masap commented:
/test pull-kubernetes-node-e2e-containerd

roycaihw reviewed (commented):
Sorry for the late reply!

Do you have numbers about how much memory was leaked and how long the memory leak lasts on etcd side? I'd expect the etcd server to have some timeout to detect a etcd client (kube-apiserver in this case) is inactive.

masap commented:
> Do you have numbers about how much memory was leaked and how long the memory leak lasts on etcd side?

It occurs 6 times during integration test test/integration/apiserver/apply.
The size of leaking data is depends on the contents.
https://github.com/kubernetes/kubernetes/blob/b1ed52195078ba5cd97bc9dc7e6b851ca22b5edb/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go#L56-L61

> I'd expect the etcd server to have some timeout to detect a etcd client (kube-apiserver in this case) is inactive.

The 6 resources are not freed during the integration test.
I checked the allocation and free with this modification.

```
diff --git a/pkg/controlplane/instance.go b/pkg/controlplane/instance.go
index a626a992d60..26da65f9a76 100644
--- a/pkg/controlplane/instance.go
+++ b/pkg/controlplane/instance.go
@@ -269,7 +269,9 @@ func (c *Config) createLeaseReconciler() reconcilers.EndpointReconciler {
 	if err != nil {
 		klog.Fatalf("Error determining service IP ranges: %v", err)
 	}
+	fmt.Println("memleaktest: createLeaseReconciler before")
 	leaseStorage, _, err := storagefactory.Create(*config, nil)
+	fmt.Println("memleaktest: createLeaseReconciler after")
 	if err != nil {
 		klog.Fatalf("Error creating storage factory: %v", err)
 	}
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
index c0fd8045f3a..ad298232416 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
@@ -229,6 +229,7 @@ func newETCD3Storage(c storagebackend.Config, newFunc func() runtime.Object) (st
 	}
 
 	client, err := newETCD3Client(c.Transport)
+	fmt.Println("memleaktest: newETCD3Storage:", client.GetId())
 	if err != nil {
 		stopCompactor()
 		return nil, nil, err
diff --git a/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go b/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go
index 2837bd4180b..96281d33635 100644
--- a/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go
+++ b/vendor/go.etcd.io/etcd/clientv3/balancer/resolver/endpoint/endpoint.go
@@ -110,6 +110,10 @@ func (e *ResolverGroup) Close() {
 	bldr.close(e.id)
 }
 
+func (e *ResolverGroup) GetId() string {
+	return e.id
+}
+
 // Build creates or reuses an etcd resolver for the etcd cluster name identified by the authority part of the target.
 func (b *builder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
 	if len(target.Authority) < 1 {
@@ -157,6 +161,7 @@ func (b *builder) close(id string) {
 	b.mu.Lock()
 	delete(b.resolverGroups, id)
 	b.mu.Unlock()
+	fmt.Println("memleaktest: close:", id)
 }
 
 func (b *builder) Scheme() string {
diff --git a/vendor/go.etcd.io/etcd/clientv3/client.go b/vendor/go.etcd.io/etcd/clientv3/client.go
index a35ec679a02..44255385450 100644
--- a/vendor/go.etcd.io/etcd/clientv3/client.go
+++ b/vendor/go.etcd.io/etcd/clientv3/client.go
@@ -180,6 +180,10 @@ func (c *Client) Sync(ctx context.Context) error {
 	return nil
 }
 
+func (c *Client) GetId() string {
+	return c.resolverGroup.GetId()
+}
+
 func (c *Client) autoSync() {
 	if c.cfg.AutoSyncInterval == time.Duration(0) {
 		return
```

masap commented on pkg/controlplane/reconcilers/lease.go:
```diff
@@ -305,5 +308,6 @@ func (r *leaseEndpointReconciler) RemoveEndpoints(serviceName string, ip net.IP,
 func (r *leaseEndpointReconciler) StopReconciling() {
 	r.reconcilingLock.Lock()
 	defer r.reconcilingLock.Unlock()
+	r.destroyFunc()
```
> Calling destroyFunc() here will cause the later Delete call racing against the etcd client destruction and cause a potential error.

It seems multi delete calls does not cause problem because the destroyFunc routine is protected by ```once```.
Ofcource I undestand we need to reduce duplicate destroy calls.
But now we need to fix a memory leak as workaroud.

https://github.com/kubernetes/kubernetes/blob/04b7977ef63247aa8e6b85fc5d12d3724ab8975c/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L243-L252
